tests/unit/completion/test_completionwidget.py

```diff
@@ -170,11 +169,36 @@ def test_completion_item_focus(which, tree, count, expected, completionview):
     assert filtermodel.data(idx) == expected
 
 
-def test_completion_item_focus_no_model(completionview):
-    """Test that next/prev won't crash with no model set.
+@pytest.mark.parametrize('show', ['always', 'auto', 'never'])
+@pytest.mark.parametrize('rows', [[], ['Aa'], ['Aa', 'Bb']])
+@pytest.mark.parametrize('quick_complete', [True, False])
+def test_completion_show(show, rows, quick_complete, completionview,
+                         config_stub):
+    """Test that the completion widget is shown at appropriate times.
 
-    This can happen if completion.show and completion.auto-open are False.
-    Regression test for issue #1722.
+    Args:
+        show: The completion show config setting.
+        rows: Each entry represents a completion category with only one item.
+        quick_complete: The completion quick-complete config setting.
     """
-    completionview.completion_item_focus('prev')
+    config_stub.data['completion']['show'] = show
+    config_stub.data['completion']['quick-complete'] = quick_complete
+
+    model = base.BaseCompletionModel()
+    for name in rows:
+        cat = QStandardItem()
+        model.appendRow(cat)
+        cat.appendRow(QStandardItem(name))
+    filtermodel = sortfilter.CompletionFilterModel(model,
+                                                   parent=completionview)
+
+    assert not completionview.isVisible()
+    completionview.set_model(filtermodel)
+    assert completionview.isVisible() == (show == 'always' and len(rows) > 0)
+    completionview.completion_item_focus('next')
+    expected = (show != 'never' and len(rows) > 0 and
```

Comment:
personally, I prefer to avoid putting much logic in the tests and instead lay out the expected results explicitly. Otherwise, your test code ends up mimicking the production code, potentially making the same mistakes and giving false positives.